### PR TITLE
Update #13 , Correct corner cases

### PR DESCRIPTION
**Pull request by rom3r4**
### Describe the change
- Add return value to copy_file() [issue 13](https://github.com/piotrmurach/tty-file/issues/13)
- Correct Source is a directory
- Correct Source does not exist
- Correct Destination is a directory

### Why are we doing this?
- Change behaviour in copy_file() method

### Benefits
- Add a return value to copy_file() - as in copy_directory()
- Treat corner-cases

### Drawbacks
-  Regressions

### Requirements
Put an X between brackets on each line if you have done the item:
[] Tests written & passing locally?
[] Code style checked?
[] Rebased with `master` branch?
[] Documentaion updated?

### CODE REVIEW

File: lib/tty/file.rb [hunk 1]
```diff
@@ -288,16 +288,26 @@ def create_file(relative_path, *args, context: nil, force: false, skip: false,
...
+      unless ::File.file? source_path
+        log_status(:error, source_path, verbose: verbose, color: :red)
+        return
+      end
+
       dest_path = (args.first || source_path).to_s.sub(/\.erb$/, "")
 
+      if ::File.directory? dest_path
+        dest_path = ::File.join(dest_path, ::File.basename(source_path))
+      end
```
Comment (piotrmurach) [inline lib/tty/file.rb:302]: Good catch! However, I think we shouldn't change the destination path. If someone wants to copy a file to an existing directory, we should allow for it and the file collision will handle the situation when the file already exists. For example,

```ruby
TTY::File.copy_file("foo/bar.rb", "baz/")
# => baz/bar.rb
```

Also, any changes like this require a test case.

File: lib/tty/file.rb [hunk 2]
```diff
@@ -288,16 +288,26 @@ def create_file(relative_path, *args, context: nil, force: false, skip: false,
     # @api public
     def copy_file(source_path, *args, context: nil, force: false, skip: false,
                   verbose: true, color: :green, noop: false, preserve: nil, &block)
+
       source_path = source_path.to_s
+      unless ::File.file? source_path
+        log_status(:error, source_path, verbose: verbose, color: :red)
+        return
+      end
```
Comment (piotrmurach) [inline lib/tty/file.rb:296]: I believe this statement makes the directory check unreachable. Also, I much prefer to allow copying a file to a directory location rather than providing an error. Wouldn't you agree?

File: lib/tty/file.rb [hunk 3]
```diff
@@ -309,10 +319,11 @@ def copy_file(source_path, *args, context: nil, force: false, skip: false,
         content = block[content] if block
         content
       end
-      return unless preserve
+      return file_dest unless preserve
 
       copy_metadata(source_path, dest_path, verbose: verbose, noop: noop,
                     color: color)
+      file_dest
```
Comment (piotrmurach) [inline lib/tty/file.rb:326]: We need a test case that confirms this behaviour so that it stays true in future releases.

File: lib/tty/file.rb [hunk 4]
```diff
@@ -309,10 +319,11 @@ def copy_file(source_path, *args, context: nil, force: false, skip: false,
         content = block[content] if block
         content
       end
-      return unless preserve
+      return file_dest unless preserve
```
Comment (piotrmurach) [inline lib/tty/file.rb:322]: We need a test case that confirms this behaviour so that it stays true in future releases.

File: lib/tty/file/create_file.rb [hunk 5]
```diff
@@ -59,6 +63,11 @@ def convert_encoded_path(filename)
       #
       # @api private
       def detect_collision
+        if directory?
+          notify(:error, :red)
+          return
+        end
```
Comment (piotrmurach) [inline lib/tty/file/create_file.rb:69]: I don't believe this is necessary if we allow copying a file to a directory destination. This should never be the case unless someone tries to copy a directory to directory, I guess we could test for that case. This would require another test case.

**Review by piotrmurach — COMMENTED** [review]
Thank you for working on this. It's looking good but before I can merge, could you please make a few changes and add tests cases to `copy_file_spec.rb` that I mentioned in my comments. For the return value, you could possibly slightly modify the existing test cases. I'd also consider adding a test case for copying a file into a directory and handle the case when the source and destination paths are both directories.
